packages/loader/container-loader/src/containerStateManager.ts

```diff
@@ -0,0 +1,208 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	ISequencedDocumentMessage,
+	ISnapshotTree,
+	IVersion,
+} from "@fluidframework/protocol-definitions";
+import { IGetPendingLocalStateProps, IRuntime } from "@fluidframework/container-definitions";
+import {
+	ITelemetryLoggerExt,
+	MonitoringContext,
+	PerformanceEvent,
+	UsageError,
+	createChildMonitoringContext,
+} from "@fluidframework/telemetry-utils";
+import { assert } from "@fluidframework/core-utils";
+import { IResolvedUrl, ISnapshot } from "@fluidframework/driver-definitions";
+import { isInstanceOfISnapshot } from "@fluidframework/driver-utils";
+import {
+	ContainerStorageAdapter,
+	ISerializableBlobContents,
+	getBlobContentsFromTree,
+} from "./containerStorageAdapter";
+import { IPendingContainerState } from "./container";
+
+export class containerStateManager {
+	private readonly savedOps: ISequencedDocumentMessage[] = [];
+	public snapshot:
+		| {
+				tree: ISnapshotTree;
+				blobs: ISerializableBlobContents;
+		  }
+		| undefined;
+	private readonly mc: MonitoringContext;
+	private readonly clientId: string | undefined;
+	private readonly offlineLoadEnabled: boolean;
+	private resolvedUrl: IResolvedUrl | undefined;
+	private runtime: IRuntime | undefined;
+	private readonly storageAdapter: ContainerStorageAdapter;
+	private readonly isInteractiveClient: boolean;
+
+	constructor(
+		subLogger: ITelemetryLoggerExt,
+		clientId,
+		offlineLoadEnabled,
+		storageAdapter,
+		isInteractiveClient,
```

Comment:
i don't really think this should be a property of this class, it make be part of the offline load computation, but doesn't need to be a property of its own.

Reply:
I don't know how to reconcile this comment with the one related to addSavedOps. I could move the condition in processRemoteMessage to containerStateManager but only if we have isInteractiveClient inside.

Reply:
there is no point is keeping isInteractiveClient  and _offlineLoadEnabled  seperate in this class, you should  include isInteractiveClient   in _offlineLoadEnabled   to simplify all the check. the current addSavedOp looks right